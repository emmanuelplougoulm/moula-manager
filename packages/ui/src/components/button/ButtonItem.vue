<template>
  <button type="button" :class="classes" :disabled="isDisabled" v-bind="$attrs" @click="onClick">
    <span v-if="hasText">
      {{ text }}
    </span>
    <component :is="icon" v-if="hasIcon" :icon-color="iconColor"></component>
  </button>
</template>

<script setup lang="ts">
import { reactive, computed } from 'vue';
import { iconValidator, iconPositionValidator, stateValidator, typeValidator } from './validator';
const innerProps = defineProps({
  icon: {
    type: String,
    validator: iconValidator,
    default: 'ArrowForwardIOS'
  },
  iconPosition: {
    type: String,
    default: 'no-icon',
    validator: iconPositionValidator
  },
  text: {
    type: String,
    default: ''
  },
  state: {
    type: String,
    validator: stateValidator,
    default: 'enabled'
  },
  type: {
    type: String,
    default: 'primary',
    validator: typeValidator
  }
});

const props = reactive(innerProps);
const classes = computed(() => ({
  [`button--${props.iconPosition}`]: props.iconPosition !== 'no-icon',
  [`button--${props.type || 'primary'}`]: true,
  ['text-small-regular']: props.type === 'link',
  ['text-big-regular']: props.type !== 'link'
}));
const iconColor = computed(() =>
  props.type === 'primary' ? 'var(--primary-text-contrast)' : 'var(--primary-text)'
);
const hasIcon = computed(() => props.iconPosition !== 'no-icon');
const hasText = computed(() => props.iconPosition !== 'icon-only');
const isDisabled = computed(() => props.state === 'disabled');
const emit = defineEmits(['click']);

function onClick() {
  emit('click');
}
</script>

<style scoped>
button {
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
  border-radius: 0.25rem;
  padding: 0.75rem 1.25rem;
  gap: 0.5rem;
  height: 3rem;
  border: none;
}

.button--primary {
  background-color: var(--primary-30);
  color: var(--primary-text-contrast);
}
.button--primary:hover {
  background-color: var(--primary-20);
}

.button--secondary {
  background-color: var(--neutral-100);
  color: var(--primary-text);
  border: 0.063rem solid var(--primary-30);
}

.button--secondary:hover {
  color: var(--primary-text-hover);
  background-color: var(--primary-90);
}

.button--icon-only {
  width: 3rem;
}

.button--left-icon {
  flex-direction: row-reverse;
}

.button--link {
  background-color: transparent;
  color: var(--primary-text);
  text-decoration: underline;
}

button:disabled {
  opacity: 0.3;
}
</style>
