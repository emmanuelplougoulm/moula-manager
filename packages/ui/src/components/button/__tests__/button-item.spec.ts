import Icons from '../../icons/index';
import { mount } from '@vue/test-utils';
import ButtonItem from '../ButtonItem.vue';

describe('Button item tests', () => {
  it('should display the button component of primary type with an icon and enabled by default', () => {
    const wrapper = mount(ButtonItem, {
      global: {
        plugins: [Icons]
      },
      props: {
        icon: 'ArrowForwardIOS',
        iconPosition: 'right-icon',
        text: 'Button',
        type: 'primary'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should display the button component of primary type and disabled', () => {
    const wrapper = mount(ButtonItem, {
      props: {
        text: 'Button',
        type: 'primary',
        state: 'disabled'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should display the button component of secondary type with an icon and enabled by default', () => {
    const wrapper = mount(ButtonItem, {
      global: {
        plugins: [Icons]
      },
      props: {
        icon: 'ArrowBackIOS',
        iconPosition: 'left-icon',
        text: 'Button',
        type: 'secondary'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should display the button component of primary type and disabled', () => {
    const wrapper = mount(ButtonItem, {
      props: {
        text: 'Button',
        type: 'secondary',
        state: 'disabled'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should display the button component of link type enabled by default', () => {
    const wrapper = mount(ButtonItem, {
      props: {
        text: 'Button',
        type: 'link'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should display the button component of link type and disabled', () => {
    const wrapper = mount(ButtonItem, {
      props: {
        text: 'Button',
        type: 'link',
        state: 'disabled'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should emit event when click on button', () => {
    const wrapper = mount(ButtonItem, {
      global: {
        plugins: [Icons]
      },
      props: {
        icon: 'ArrowForwardIOS',
        iconPosition: 'right-icon',
        text: 'Button',
        type: 'primary'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.find('button').trigger('click');
    expect(wrapper.emitted()).toHaveProperty('click');
  });
});
