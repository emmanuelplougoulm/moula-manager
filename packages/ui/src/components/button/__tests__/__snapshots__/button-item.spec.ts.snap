// Vitest Snapshot v1

exports[`Button item tests > should display the button component of link type and disabled 1`] = `
"<button data-v-71b6af39=\\"\\" type=\\"button\\" class=\\"button--link text-small-regular\\" disabled=\\"\\"><span data-v-71b6af39=\\"\\">Button</span>
  <!--v-if-->
</button>"
`;

exports[`Button item tests > should display the button component of link type enabled by default 1`] = `
"<button data-v-71b6af39=\\"\\" type=\\"button\\" class=\\"button--link text-small-regular\\"><span data-v-71b6af39=\\"\\">Button</span>
  <!--v-if-->
</button>"
`;

exports[`Button item tests > should display the button component of primary type and disabled 1`] = `
"<button data-v-71b6af39=\\"\\" type=\\"button\\" class=\\"button--primary text-big-regular\\" disabled=\\"\\"><span data-v-71b6af39=\\"\\">Button</span>
  <!--v-if-->
</button>"
`;

exports[`Button item tests > should display the button component of primary type and disabled 2`] = `
"<button data-v-71b6af39=\\"\\" type=\\"button\\" class=\\"button--secondary text-big-regular\\" disabled=\\"\\"><span data-v-71b6af39=\\"\\">Button</span>
  <!--v-if-->
</button>"
`;

exports[`Button item tests > should display the button component of primary type with an icon and enabled by default 1`] = `
"<button data-v-71b6af39=\\"\\" type=\\"button\\" class=\\"button--right-icon button--primary text-big-regular\\"><span data-v-71b6af39=\\"\\">Button</span>
  <div data-v-b9dd0015=\\"\\" data-v-71b6af39=\\"\\" class=\\"container\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
      <path d=\\"M304 987.566 234.434 917l343-343-343-343L304 160.434 717.566 574 304 987.566Z\\"></path>
    </svg></div>
</button>"
`;

exports[`Button item tests > should display the button component of secondary type with an icon and enabled by default 1`] = `
"<button data-v-71b6af39=\\"\\" type=\\"button\\" class=\\"button--left-icon button--secondary text-big-regular\\"><span data-v-71b6af39=\\"\\">Button</span>
  <div data-v-b9dd0015=\\"\\" data-v-71b6af39=\\"\\" class=\\"container\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
      <path d=\\"M656.13 989.566 241.999 576 656.13 162.434 725.696 233l-343 343 343 343-69.566 70.566Z\\"></path>
    </svg></div>
</button>"
`;

exports[`Button item tests > should emit event when click on button 1`] = `
"<button data-v-71b6af39=\\"\\" type=\\"button\\" class=\\"button--right-icon button--primary text-big-regular\\"><span data-v-71b6af39=\\"\\">Button</span>
  <div data-v-b9dd0015=\\"\\" data-v-71b6af39=\\"\\" class=\\"container\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
      <path d=\\"M304 987.566 234.434 917l343-343-343-343L304 160.434 717.566 574 304 987.566Z\\"></path>
    </svg></div>
</button>"
`;
